lib/api/core/hotelClerk.js

```diff
@@ -191,445 +184,406 @@ function HotelClerk (kuzzle) {
+          controller: 'document',
+          action: 'search',
+          index,
+          collection
+        }, request.context);
+
+        promises.push(request.context.user.isActionAllowed(isAllowedRequest, this.kuzzle)
           .then(isAllowed => {
             if (!isAllowed) {
-              return callback(null);
+              return;
             }
 
-            if (!list[room.index]) {
-              list[room.index] = {};
+            if (!list[index]) {
+              list[index] = {};
+            }
+            if (!list[index][collection]) {
+              list[index][collection] = {};
             }
 
-            if (!list[room.index][room.collection]) {
-              list[room.index][room.collection] = {};
+            for (const roomId of dslIndex[index][collection]) {
+              const room = this.rooms[roomId];
+              // the room may be currently registering in the dsl and not ready in the hotel clerk
+              // or already deleted in from the hotel clerk and still in the dsl
+              if (room) {
+                list[index][collection][roomId] = room.customers.size;
+              }
             }
+          }));
+      }
+    }
 
-            list[room.index][room.collection][roomId] = room.customers.length;
-            callback(null);
-          });
-      }, () => resolve(list));
-    });
-  };
+    return Bluebird.all(promises)
+      .then(() => list);
+  }
 
   /**
-   * Joins an existing room.
+   * This function will delete a user from this.customers, and
+   * decrement the subscribers count in all rooms where he has subscribed to
+   * Call the cleanUpRooms function to manage empty rooms
+   * Usually called on a user disconnection event
    *
-   * @param {Request} request
-   * @returns {Promise}
+   * @param {RequestContext} requestContext
    */
-  this.join = function hotelClerkJoin (request) {
-    const roomId = request.input.body.roomId;
-
-    if (!this.rooms[roomId]) {
-      return Bluebird.reject(new NotFoundError('No room found for id ' + roomId));
+  removeCustomerFromAllRooms (requestContext) {
+    if (!this.customers[requestContext.connectionId]) {
+      // No need to raise an error if the connection has already been cleaned up
+      return;
     }
 
-    return Bluebird.resolve(subscribeToRoom.call(this, roomId, request))
-      .then(response => {
-        kuzzle.pluginsManager.trigger('core:hotelClerk:join', response);
-        return response.data;
-      });
-  };
+    for (const roomId of Object.keys(this.customers[requestContext.connectionId])) {
+      try {
+        this._removeRoomForCustomer(requestContext, roomId);
+      }
+      catch (err) {
+        this.kuzzle.pluginsManager.trigger('log:error', err);
+      }
+    }
+  }
 
   /**
    * Remove rooms for a given collection
    * If rooms attribute is not provided, all rooms for the collection are removed
    * @param {Request} request
-   * @returns {Promise}
+   * @return {Promise<Object>}
```

Comment:
The first `send` is called on Kuzzle start.
The next one is attached to the `onConnectHandlers` of the proxy ws client and is triggered when reconnecting to the proxy, for instance in case of a network error.